src/Oscoin/Crypto/Hash/RealWorld.hs

```diff
@@ -0,0 +1,125 @@
+
+import           Codec.Serialise
+import qualified Codec.Serialise.Multihash as Multihash.CBOR
+import           Control.Monad.Fail (fail)
+import qualified Crypto.Data.Auth.Tree.Cryptonite as Cryptonite
+import           Crypto.Data.Auth.Tree.Internal (MerkleHash(..))
+import           Crypto.Hash (Blake2b_256(..), Digest)
+import qualified Crypto.Hash as Crypto
+import           Data.Aeson (FromJSON(..), withText)
+import           Data.ByteArray (ByteArrayAccess, convert)
+import qualified Data.ByteArray as ByteArray
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.BaseN as BaseN
+import qualified Data.ByteString.Lazy as LBS
+import           Data.Maybe (fromJust)
+import           Data.Multihash (Multihashable)
+import qualified Data.Text as T
+import qualified Database.SQLite.Simple as Sql
+import qualified Database.SQLite.Simple.FromField as Sql
+import qualified Database.SQLite.Simple.Ok as Sql
+import           Formatting.Buildable (Buildable(..))
+import           Oscoin.Crypto (RealWorldCrypto)
+import           Oscoin.Crypto.Hash
+import           Web.HttpApiData (FromHttpApiData(..))
+
+{------------------------------------------------------------------------------
+  Crypto-specific instances
+------------------------------------------------------------------------------}
+
+instance HasHashing RealWorldCrypto where
+    type HashAlgorithm RealWorldCrypto = Blake2b_256
+
+    -- | A 'Digest' obtained by some hash algorithm @(HashAlgorithm RealWorldCrypto)@.
+    --
+    -- Textual representations use base-58 encoding. Serialisation and
+    -- deserialisation via "Crypto.Hash.Multi".
+    newtype Hash RealWorldCrypto =
+        Hash { fromHash :: Digest (HashAlgorithm RealWorldCrypto) }
+        deriving (Eq, Ord, ByteArrayAccess)
+
+    fromByteArray = Hash . Crypto.hash
+    toByteArray = convert . fromHash
+
+    hashAlgorithm = Blake2b_256
+
+    zeroHash = Hash . fromJust $
+        Crypto.digestFromByteString @(HashAlgorithm RealWorldCrypto) @ByteString
+            (ByteArray.zero (hashDigestSize (Proxy @(HashAlgorithm RealWorldCrypto))))
+
+{------------------------------------------------------------------------------
+  Instances galore
+-------------------------------------------------------------------------------}
+
+instance Hashable RealWorldCrypto Text where
+    hash = toHashed . Hash . Crypto.hash . encodeUtf8
+
+instance Hashable RealWorldCrypto ByteString where
+    hash = toHashed . Hash . Crypto.hash
+
+instance Hashable RealWorldCrypto LByteString where
+    hash = toHashed . fromHashed . hash . LBS.toStrict
+
+instance Hashable RealWorldCrypto Word8 where
+    hash = toHashed . Hash . Crypto.hash . BS.singleton
+
+instance (ByteArrayAccess a) => Hashable RealWorldCrypto (Maybe a) where
+    hash (Just x) = toHashed . Hash $ Crypto.hash x
+    hash Nothing  = toHashed $ zeroHash
+
+instance Multihashable (HashAlgorithm RealWorldCrypto) => FromJSON (Hash RealWorldCrypto) where
```

Comment:
The `Multihashable` constraints here can be omitted (not smaller than instance head), and `UndecidableInstances` can be removed.

Reply:
Something that GHC/stack/ghcid performs (seemingly) arbitrarily is exactly warning about redundant constraints. It seems like that gets triggered only if you modify a module deep down the guts which causes (forces) a recompile. Isn't there a more aggressive compiler flag to do this all the time (unlikely, but still)

Reply:
I think this doesn't trigger a warning because `UndecidableInstances`

Reply:
Comment still stands :)

Reply:
Thanks for the reminder! You were right, the constraint was unnecessary but alas we still need `UndecidableInstances` for the `Serialise` instance in this module. Oh well 😉